### ISSUE TITLE
OOM issues and Sequence Lengths

### ISSUE THREAD

**Issue by twuebi** [stickeritis/sticker#55]
Currently it is impossible to train moderately sized transformers on e.g. TüBa-D/Z since the total VRAM consumption is governed by the longest sequence in the train set.

Possible solutions:

1. introduce an optional maximum length for training sentences. In TüBa, there are for instance only few sentences with more than 150 words while the longest has 234 words. Dropping them may not be very hurtful.

2. I did some experiments with dynamic batch sizes by setting a capacity to e.g. 20000 words. In that case a batch could consist of 200 sentences with the longest one having 100 words, 100 sentences with 200 etc. I did not observe big differences in performance, but did not do exhaustive experiments. I'd believe that there may be some interactions with batch normalization.

**Comment by danieldk** [comment]
Solution 1 was implemented by #88. I think we can close this for now, right?

(There may be more sophisticated solutions, but like you said elsewhere, most sentences above 100-150 words are typically not real sentences.)

**Comment by twuebi** [comment]
> Solution 1 was implemented by #88. I think we can close this for now, right?

Sure.



> (There may be more sophisticated solutions, but like you said elsewhere, most sentences above 100-150 words are typically not real sentences.)

There are also interactions with learning rate and batch size. AFAIR, in some (google?) paper or maybe the tensor2tensor library, they increase batch size instead of lr decay.
